src/transport/http2.rs

```diff
@@ -294,49 +323,41 @@ fn try_get_length_prefixed(buf: &[u8]) -> Option<&[u8]> {
 ///
 /// The HTTP2 request comes from *req* and the data is sent to `req_tx`.
 fn spawn_recv_forwarder<In: RpcMessage>(
-    req: Body,
+    req: impl Stream<Item = result::Result<Bytes, hyper::Error>> + Send + Unpin + 'static,
     req_tx: Sender<result::Result<In, RecvError>>,
-) -> JoinHandle<()> {
+) -> JoinHandle<result::Result<(), ()>> {
     tokio::spawn(async move {
         let mut stream = req;
         let mut buf = Vec::new();
 
         while let Some(chunk) = stream.next().await {
-            let exit = match chunk.as_ref() {
+            match chunk.as_ref() {
                 Ok(chunk) => {
                     event!(Level::TRACE, "Server got {} bytes", chunk.len());
-                    buf.extend_from_slice(chunk);
-                    false
+                    if buf.is_empty() {
+                        // try to forward directly from buffer
+                        let sent = try_forward_all(chunk, &req_tx).await?;
+                        // add just the rest, if any
+                        buf.extend_from_slice(&chunk[sent..]);
+                    } else {
+                        // no choice but to add it all
+                        buf.extend_from_slice(chunk);
+                    }
                 }
                 Err(cause) => {
                     // Indicates that the connection has been closed on the client side.
                     // This is a normal occurrence, e.g. when the client has raced the RPC
                     // call with something else and has droppped the future.
                     debug!("Network error: {}", cause);
-                    true
-                }
-            };
-            // Drain all complete messages from the buffer, even in case of an error.
-            while let Some(msg) = try_get_length_prefixed(&buf) {
-                let item = bincode::deserialize(msg).map_err(RecvError::DeserializeError);
-                buf.drain(..4 + msg.len());
-                if let Err(_cause) = req_tx.send_async(item).await {
-                    // The receiver is gone, so we can't send any more data.
-                    //
-                    // This is a normal way for an interaction to end, when the server side is done processing
-                    // the request and drops the receiver.
-                    //
-                    // don't log the cause. It does not contain any useful information.
-                    trace!("Flume receiver dropped");
                     break;
                 }
-            }
-            // exiting the task will drop the sender, which will cause the receiver to produce an error.
-            // this is a normal way for an interaction to end.
-            if exit {
-                break;
-            }
+            };
+            let sent = try_forward_all(&buf, &req_tx).await?;
```

Comment:
I hope rust / llvm will be smart enough to figure this out. try_forward_all calls try_get_length_prefixed, which will immediately return None if the buffer is < 4 bytes.

Given all the other stuff that is going on, I would guess that it would be hard to measure the difference. I may be wrong, just my hunch.